### PR TITLE
8289692: JFR: Thread checkpoint no longer enforce mutual exclusion post Loom integration

### PR DESCRIPTION
**Pull request by mgronlun**
Greetings,

Before integrating Loom and Virtual Threads, threads writing checkpoint meta-information claimed a checkpoint buffer from a list categorized as thread-local.
Buffers on this list are not reset eagerly as part of a flushpoint but only as part of chunk rotation, where the JFR Recorder Thread has exclusive access by issuing a prior epoch shift.

With Loom, the thread-local buffers are exclusively dedicated to meta-information for Virtual Threads.
The meta-information about the JVM thread, for example, the carrier thread, is written to a buffer located on the list categorized as global. Buffers on this list are reset eagerly during flushpoints. Before Loom, an invariant for this global list was that only the JFR Recorder Thread used it.
This invariant no longer holds. The JFR Recorder Thread will reset buffers on this list without protection and can do so in the middle of another thread's attempted writes. Some consequences are asserts in debug builds and data loss in product builds.

Some mutex mechanism for the buffers on the global list must be re-introduced to prevent the JFR Recorder Thread from resetting buffers currently in use. This change set adds back proper protection.

Testing: jdk_jfr, stress

Thanks
Markus

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289692](https://bugs.openjdk.org/browse/JDK-8289692): JFR: Thread checkpoint no longer enforce mutual exclusion post Loom integration

### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [dd40dbb5](https://git.openjdk.org/jdk19/pull/109/files/dd40dbb50c07b1da38463dbf7ad4b769e55c6c7c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk19 pull/109/head:pull/109` \
`$ git checkout pull/109`

Update a local copy of the PR: \
`$ git checkout pull/109` \
`$ git pull https://git.openjdk.org/jdk19 pull/109/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 109`

View PR using the GUI difftool: \
`$ git pr show -t 109`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk19/pull/109.diff">https://git.openjdk.org/jdk19/pull/109.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/jfr/recorder/storage/jfrStorageUtils.inline.hpp [hunk 1]
```diff
@@ -91,24 +91,27 @@ inline bool MutexedWriteOp<Operation>::process(typename Operation::Type* t) {
 }
 
 template <typename Type>
-static void retired_sensitive_acquire(Type* t) {
+static void retired_sensitive_acquire(Type* t, Thread* thread) {
   assert(t != NULL, "invariant");
+  assert(thread != nullptr, "invariant");
```
Comment (robehn) [inline src/hotspot/share/jfr/recorder/storage/jfrStorageUtils.inline.hpp:96]: Maybe add assertion that thread is current.

**Review by robehn — APPROVED** [review]
Thank, looks good.

**Comment by mgronlun** [comment]
Thank you @robehn for reviewing!

**Pull request closed without merging**